man/msgsed.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.24.
.TH LT-MSGSED "1" "June 2001" "GNU gettext 0.11" GNU
.SH NAME
msgsed \- stream editing of message catalog
.SH SYNOPSIS
.B lt-msgsed
[\fIOPTION\fR] [\fIINPUTFILE\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Applies a sed script to all translations of a translation catalog.
.PP
Mandatory arguments to long options are mandatory for short options too.
.SS "Input file location:"
.TP
INPUTFILE
input PO file
.TP
\fB\-D\fR, \fB\-\-directory\fR=\fIDIRECTORY\fR
add DIRECTORY to list for input files search
.PP
If no input file is given or if it is -, standard input is read.
.SS "Output file location:"
.TP
\fB\-o\fR, \fB\-\-output\-file\fR=\fIFILE\fR
write output to specified file
.PP
The results are written to standard output if no output file is specified
or if it is -.
.SS "Sed options:"
.TP
\fB\-e\fR, \fB\-\-expression\fR=\fISCRIPT\fR
add SCRIPT to the commands to be executed
.TP
\fB\-f\fR, \fB\-\-file\fR=\fISCRIPTFILE\fR
add the contents of SCRIPTFILE to the commands
to be executed
.TP
\fB\-n\fR, \fB\-\-quiet\fR, \fB\-\-silent\fR
suppress automatic printing of pattern space
.SS "Output details:"
.TP
\fB\-\-no\-escape\fR
do not use C escapes in output (default)
.TP
\fB\-E\fR, \fB\-\-escape\fR
use C escapes in output, no extended chars
.TP
\fB\-\-force\-po\fR
write PO file even if empty
.TP
\fB\-i\fR, \fB\-\-indent\fR
indented output style
.TP
\fB\-\-no\-location\fR
suppress '#: filename:line' lines
.TP
\fB\-\-add\-location\fR
preserve '#: filename:line' lines (default)
.TP
\fB\-\-strict\fR
strict Uniforum output style
.TP
\fB\-w\fR, \fB\-\-width\fR=\fINUMBER\fR
set output page width
.TP
\fB\-s\fR, \fB\-\-sort\-output\fR
generate sorted output and remove duplicates
.TP
\fB\-F\fR, \fB\-\-sort\-by\-file\fR
sort output by file location
.SS "Informative output:"
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
output version information and exit
.SH AUTHOR
Written by Bruno Haible.
.SH "REPORTING BUGS"
Report bugs to <bug-gnu-utils@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2001 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
The full documentation for
.B lt-msgsed
is maintained as a Texinfo manual.  If the
.B info
and
.B lt-msgsed
programs are properly installed at your site, the command
.IP
.B info lt-msgsed
.PP
should give you access to the complete manual.


NEWS
Version 1.1 - October 2026

# Programming languages support:
  * C++, Rust:
    - The --keyword option now supports "qualified names" of the form
      'namespace::function', that specify a function in a namespace or
      possibly even a function in a nested namespace.
  * Scheme:
    - A new example 'hello-scheme' has been added.

Version 1.0 - January 2026

# Improvements for maintainers and distributors:
  * In a po/ directory, the PO files are now exactly those that the
    translators submitted or committed in version control, or a
    translation project's daemon committed on behalf of the translators.
    They are no longer regularly updated with respect to the POT file
    in the same directory.

    The advantage for maintainers is that the maintainer may commit the
    PO files in version control, without getting
      - lots of modified files shown by "git status",
      - frequent merge conflicts when merging between branches,
      - a voluminous version control history.

    The advantage for distributors is that the role of files in a
    release tarball are clearer: The PO files are source code, whereas
    the POT file and the *.gmo files are generated files.

    ATTENTION translators!
    Translators who work directly on a package's source code (without
    going through a translation project) now need to run "msginit"
    before starting work on a PO file.

  * A new program 'po-fetch' is provided, that fetches the translated
    PO files from a translation project's site on the internet, and
    updates the LINGUAS file accordingly.

  * In a po/ directory, a new script 'fetch-po' is now added by 'gettextize'.
    It provides the standard interface for fetching the translated PO files.
    It typically either invokes the 'po-fetch' program or does nothing.

# Improvements for translators:
  * msginit:
    - When the PO file already exists, 'msginit' now updates it w.r.t. the
      POT file, like 'msgmerge' would do. Previously, 'msginit' failed with
      an error message in this situation.

  * Pretranslation:
    - Two new programs, 'msgpre' and 'spit', are provided, that implement
      machine translation through a locally installed Large Language Model
      (LLM).  'msgpre' applies to an entire PO file, 'spit' to a single
      message.
    - The documentation has a new chapter "Pretranslation".

# Improvements for maintainers:
  * xgettext:
    - The refactoring suggestion when a translatable string contains an URL
      or email address can now be inhibited through a command-line option
      '--no-check=url' or '--no-check=email', or through a comment in the
      source code of the form
          /* xgettext: no-url-check */
      or
          /* xgettext: no-email-check */

# Programming languages support:
  * OCaml:
    - xgettext now supports OCaml.
    - 'msgfmt -c' now verifies the syntax of translations of OCaml format
      strings.
    - A new example 'hello-ocaml' has been added.
  * Rust:
    - xgettext now recognizes 'gettextrs::gettext' invocations, like 'gettext'
      invocations.

# libgettextpo library:
  * The function 'po_message_get_format' now supports distinguishing whether
    a negative format string mark, such as 'no-c-format', is set or not.
  * The new functions
      po_message_has_workflow_flag
      po_message_set_workflow_flag
      po_message_workflow_flags_iterator, po_flag_next, po_flag_iterator_free
    can be used to manipulate or inspect the workflow flags of a message.
  * The new functions
      po_message_has_sticky_flag
      po_message_set_sticky_flag
      po_message_sticky_flags_iterator, po_flag_next, po_flag_iterator_free
    can be used to manipulate or inspect the sticky flags of a messsage.

# Emacs PO mode:
  Restore syntax highlighting in Emacs version 30 or newer.

Version 0.26 - July 2025

# Programming languages support:
  * JavaScript:
    - xgettext now parses regular expressions with character classes
      correctly.
  * C, C++, Python, JavaScript, EmacsLisp, librep, Go, Ruby, awk, D, Tcl,
    Perl, PHP:
    - xgettext's heuristic recognition of format strings has been improved:
      strings like "100% complete" (with a space flag in a format directive)
      are no longer flagged as format strings by default, unless they occur
      in a context that requires a format string.  You can override this
      heuristic by using a comment of the form /* xgettext: c-format */.
  * Shell:
    - The documentation now mentions two other approaches for
      internationalizing messages with parameters in shell scripts.
    - xgettext now recognizes format strings in the 'printf' command syntax.
      They are marked as 'sh-printf-format' in POT and PO files.
    - Two new programs 'printf_gettext' and 'printf_ngettext' are provided,
      that do formatted output with a localized format string in a more
      efficient way (without spawning a subshell).
    - xgettext now recognizes the \c, \u, and \U escape sequences in dollar-
      single-quoted strings $'...'.

# Improvements for maintainers:
  * xgettext:
    - When extracting a message with plural that is some format string,
      xgettext now verifies that the msgid and msgid_plural are compatible
      as format strings.  For most format string types, this still allows
      omitting from msgid a placeholder that is used in msgid_plural.  But
      when a placeholder is used in both msgid and msgid_plural, its type
      must be the same in both.
    - xgettext now suggests a refactoring when a translatable string
      contains an URL or email address.

# Improvements for translators:
  * msggrep:
    - msggrep accepts two new options -W/--workflow-flags and -S/--sticky-flags
      that allow to select only messages that have a specified flag.

# Bug fixes:
  - The AM_GNU_GETTEXT macro now rejects the dysfunctional gettext() function
    in libc of Solaris 11.[0-3], Solaris OpenIndiana, and Solaris OmniOS.

src/msgsed.c
/* Edit translations using a script of editing commands.
   Copyright (C) 2001 Free Software Foundation, Inc.
   Written by Bruno Haible <haible@clisp.cons.org>, 2001.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */


#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/time.h>

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif

#ifdef HAVE_LIMITS_H
# include <limits.h>
#endif

#include "dir-list.h"
#include "error.h"
#include "progname.h"
#include "message.h"
#include "read-po.h"
#include "write-po.h"
#include "str-list.h"
#include "msgl-charset.h"
#include "system.h"
#include "findprog.h"
#include "pipe.h"
#include "wait-process.h"
#include "libgettext.h"

#define _(str) gettext (str)


/* We use a child process running 'sed', and communicate through a
   bidirectional pipe.  To avoid deadlocks, let the child process decide
   when it wants to read or to write, and let the parent behave accordingly.
   The parent uses select() to know whether it must write or read.  On
   platforms without select(), we use non-blocking I/O.  (This means the
   parent is busy looping while waiting for the child.  Not good.)  */

/* On BeOS select() works only on sockets, not on normal file descriptors.  */
#ifdef __BeOS__
# undef HAVE_SELECT
#endif


/* Force output of PO file even if empty.  */
static int force_po;

/* Arguments to be passed to the sed subprocess.  */
static string_list_ty *sed_args;

/* Pathname of the sed program.  */
static const char *sed_path;

/* Argument list for the sed program.  */
static char **sed_argv;

/* Long options.  */
static const struct option long_options[] =
{
  { "add-location", no_argument, &line_comment, 1 },
  { "directory", required_argument, NULL, 'D' },
  { "escape", no_argument, NULL, 'E' },
  { "expression", required_argument, NULL, 'e' },
  { "file", required_argument, NULL, 'f' },
  { "force-po", no_argument, &force_po, 1 },
  { "help", no_argument, NULL, 'h' },
  { "indent", no_argument, NULL, 'i' },
  { "no-escape", no_argument, NULL, CHAR_MAX + 1 },
  { "no-location", no_argument, &line_comment, 0 },
  { "output-file", required_argument, NULL, 'o' },
  { "quiet", no_argument, NULL, 'n' },
  { "silent", no_argument, NULL, 'n' },
  { "sort-by-file", no_argument, NULL, 'F' },
  { "sort-output", no_argument, NULL, 's' },
  { "strict", no_argument, NULL, 'S' },
  { "version", no_argument, NULL, 'V' },
  { "width", required_argument, NULL, 'w', },
  { NULL, 0, NULL, 0 }
};


/* Prototypes for local functions.  */
static void usage PARAMS ((int status));
#ifdef EINTR
static inline int nonintr_close PARAMS ((int fd));
static inline ssize_t nonintr_read PARAMS ((int fd, void *buf, size_t count));
static inline ssize_t nonintr_write PARAMS ((int fd, const void *buf,
					     size_t count));
#endif
static void process_string PARAMS ((const char *str, size_t len,
				    char **resultp, size_t *lengthp));
static void process_message PARAMS ((message_ty *mp));
static void process_message_list PARAMS ((message_list_ty *mlp));
static msgdomain_list_ty *
       process_msgdomain_list PARAMS ((msgdomain_list_ty *mdlp));


int
main (argc, argv)
     int argc;
     char **argv;
{
  int opt;
  bool do_help;
  bool do_version;
  char *output_file;
  const char *input_file;
  msgdomain_list_ty *result;
  bool sort_by_filepos = false;
  bool sort_by_msgid = false;
  size_t i;

  /* Set program name for messages.  */
  program_name = argv[0];
  error_print_progname = maybe_print_progname;

#ifdef HAVE_SETLOCALE
  /* Set locale via LC_ALL.  */
  setlocale (LC_ALL, "");
#endif

  /* Set the text message domain.  */
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  /* Set default values for variables.  */
  do_help = false;
  do_version = false;
  output_file = NULL;
  input_file = NULL;
  sed_args = string_list_alloc ();

  while ((opt = getopt_long (argc, argv, "D:e:Ef:Fhino:sVw:",
			     long_options, NULL))
	 != EOF)
    switch (opt)
      {
      case '\0':		/* Long option.  */
	break;

      case 'D':
	dir_list_append (optarg);
	break;

      case 'e':
	string_list_append (sed_args, "-e");
	string_list_append (sed_args, optarg);
	break;

      case 'E':
	message_print_style_escape (true);
	break;

      case 'f':
	string_list_append (sed_args, "-f");
	string_list_append (sed_args, optarg);
	break;

      case 'F':
	sort_by_filepos = true;
	break;

      case 'h':
	do_help = true;
	break;

      case 'i':
	message_print_style_indent ();
	break;

      case 'n':
	string_list_append (sed_args, "-n");
	break;

      case 'o':
	output_file = optarg;
	break;

      case 's':
	sort_by_msgid = true;
	break;

      case 'S':
	message_print_style_uniforum ();
	break;

      case 'V':
	do_version = true;
	break;

      case 'w':
	{
	  int value;
	  char *endp;
	  value = strtol (optarg, &endp, 10);
	  if (endp != optarg)
	    message_page_width_set (value);
	}
	break;

      case CHAR_MAX + 1:
	message_print_style_escape (false);
	break;

      default:
	usage (EXIT_FAILURE);
	break;
      }

  /* Version information is requested.  */
  if (do_version)
    {
      printf ("%s (GNU %s) %s\n", basename (program_name), PACKAGE, VERSION);
      /* xgettext: no-wrap */
      printf (_("Copyright (C) %s Free Software Foundation, Inc.\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\
"),
	      "2001");
      printf (_("Written by %s.\n"), "Bruno Haible");
      exit (EXIT_SUCCESS);
    }

  /* Help is requested.  */
  if (do_help)
    usage (EXIT_SUCCESS);

  /* Test whether we have an .po file name as argument.  */
  if (optind == argc)
    input_file = "-";
  else if (optind + 1 == argc)
    input_file = argv[optind];
  else
    {
      error (EXIT_SUCCESS, 0, _("at most one input file allowed"));
      usage (EXIT_FAILURE);
    }

  if (sed_args->nitems == 0)
    error (EXIT_FAILURE, 0, _("at least one sed script must be specified"));

  if (sort_by_msgid && sort_by_filepos)
    error (EXIT_FAILURE, 0, _("%s and %s are mutually exclusive"),
	   "--sort-output", "--sort-by-file");

  /* Read input file.  */
  result = read_po_file (input_file);

  /* Warn if the current locale is not suitable for this PO file.  */
  compare_po_locale_charsets (result);

  /* Attempt to locate the 'sed' program.
     This is an optimization, to avoid that spawn/exec searches the PATH
     on every call.  */
  sed_path = find_in_path ("sed");

  /* Build argument list for the 'sed' program.  */
  sed_argv = (char **) xmalloc ((1 + sed_args->nitems + 1) * sizeof (char *));
  sed_argv[0] = (char *) sed_path;
  for (i = 1; i <= sed_args->nitems; i++)
    sed_argv[i] = (char *) sed_args->item[i - 1];
  sed_argv[i] = NULL;

  /* Apply the sed script.  */
  result = process_msgdomain_list (result);

  /* Sort the results.  */
  if (sort_by_filepos)
    msgdomain_list_sort_by_filepos (result);
  else if (sort_by_msgid)
    msgdomain_list_sort_by_msgid (result);

  /* Write the merged message list out.  */
  msgdomain_list_print (result, output_file, force_po, false);

  exit (EXIT_SUCCESS);
}


/* Display usage information and exit.  */
static void
usage (status)
     int status;
{
  if (status != EXIT_SUCCESS)
    fprintf (stderr, _("Try `%s --help' for more information.\n"),
	     program_name);
  else
    {
      /* xgettext: no-wrap */
      printf (_("\
Usage: %s [OPTION] [INPUTFILE]\n\
"), program_name);
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Applies a sed script to all translations of a translation catalog.\n\
"));
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Mandatory arguments to long options are mandatory for short options too.\n\
"));
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Input file location:\n\
  INPUTFILE                   input PO file\n\
  -D, --directory=DIRECTORY   add DIRECTORY to list for input files search\n\
If no input file is given or if it is -, standard input is read.\n\
"));
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Output file location:\n\
  -o, --output-file=FILE      write output to specified file\n\
The results are written to standard output if no output file is specified\n\
or if it is -.\n\
"));
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Sed options:\n\
  -e, --expression=SCRIPT     add SCRIPT to the commands to be executed\n\
  -f, --file=SCRIPTFILE       add the contents of SCRIPTFILE to the commands\n\
                                to be executed\n\
  -n, --quiet, --silent       suppress automatic printing of pattern space\n\
"));
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Output details:\n\
      --no-escape             do not use C escapes in output (default)\n\
  -E, --escape                use C escapes in output, no extended chars\n\
      --force-po              write PO file even if empty\n\
  -i, --indent                indented output style\n\
      --no-location           suppress '#: filename:line' lines\n\
      --add-location          preserve '#: filename:line' lines (default)\n\
      --strict                strict Uniforum output style\n\
  -w, --width=NUMBER          set output page width\n\
  -s, --sort-output           generate sorted output and remove duplicates\n\
  -F, --sort-by-file          sort output by file location\n\
"));
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Informative output:\n\
  -h, --help                  display this help and exit\n\
  -V, --version               output version information and exit\n\
"));
      printf ("\n");
      fputs (_("Report bugs to <bug-gnu-utils@gnu.org>.\n"),
	     stdout);
    }

  exit (status);
}


#ifdef EINTR

/* EINTR handling for close(), read(), write(), select().
   These functions can return -1/EINTR even though we don't have any
   signal handlers set up, namely when we get interrupted via SIGSTOP.  */

static inline int
nonintr_close (fd)
     int fd;
{
  int retval;

  do
    retval = close (fd);
  while (retval < 0 && errno == EINTR);

  return retval;
}
#define close nonintr_close

static inline ssize_t
nonintr_read (fd, buf, count)
     int fd;
     void *buf;
     size_t count;
{
  ssize_t retval;

  do
    retval = read (fd, buf, count);
  while (retval < 0 && errno == EINTR);

  return retval;
}
#define read nonintr_read

static inline ssize_t
nonintr_write (fd, buf, count)
     int fd;
     const void *buf;
     size_t count;
{
  ssize_t retval;

  do
    retval = write (fd, buf, count);
  while (retval < 0 && errno == EINTR);

  return retval;
}
#define write nonintr_write

# if HAVE_SELECT

static inline int
nonintr_select (n, readfds, writefds, exceptfds, timeout)
     int n;
     fd_set *readfds;
     fd_set *writefds;
     fd_set *exceptfds;
     struct timeval *timeout;
{
  int retval;

  do
    retval = select (n, readfds, writefds, exceptfds, timeout);
  while (retval < 0 && errno == EINTR);

  return retval;
}
#define select nonintr_select

# endif

#endif


/* Non-blocking I/O.  */
#ifndef O_NONBLOCK
# define O_NONBLOCK O_NDELAY
#endif
#if HAVE_SELECT
# define IS_EAGAIN(errcode) 0
#else
# ifdef EWOULDBLOCK
#  define IS_EAGAIN(errcode) ((errcode) == EAGAIN || (errcode) == EWOULDBLOCK)
# else
#  define IS_EAGAIN(errcode) ((errcode) == EAGAIN)
# endif
#endif

/* Process a string STR of size LEN bytes through sed, then remove NUL bytes.
   Store the freshly allocated result at *RESULTP and its length at *LENGTHP.
 */
static void
process_string (str, len, resultp, lengthp)
     const char *str;
     size_t len;
     char **resultp;
     size_t *lengthp;
{
  pid_t child;
  int fd[2];
  char *result;
  size_t allocated;
  size_t length;
  int exitstatus;

  /* Open a bidirectional pipe to a sed subprocess.  */
  child = create_pipe_bidi ("sed", sed_path, sed_argv, fd);
     
  /* Enable non-blocking I/O.  This permits the read() and write() calls
     to return -1/EAGAIN without blocking; this is important for polling
     if HAVE_SELECT is not defined.  It also permits the read() and write()
     calls to return after partial reads/writes; this is important if
     HAVE_SELECT is defined, because select() only says that some data
     can be read or written, not how many.  Without non-blocking I/O,
     Linux 2.2.17 and BSD systems prefer to block instead of returning
     with partial results.  */
  {
    int fcntl_flags;

    if ((fcntl_flags = fcntl (fd[1], F_GETFL, 0)) < 0
	|| fcntl (fd[1], F_SETFL, fcntl_flags | O_NONBLOCK) < 0
	|| (fcntl_flags = fcntl (fd[0], F_GETFL, 0)) < 0
	|| fcntl (fd[0], F_SETFL, fcntl_flags | O_NONBLOCK) < 0)
      error (EXIT_FAILURE, errno,
	     _("cannot set up nonblocking I/O to sed subprocess"));
  }

  allocated = len + (len >> 2) + 1;
  result = (char *) xmalloc (allocated);
  length = 0;

  for (;;)
    {
#if HAVE_SELECT
      int n;
      fd_set readfds;
      fd_set writefds;

      FD_ZERO (&readfds);
      FD_SET (fd[0], &readfds);
      n = fd[0] + 1;
      if (str != NULL)
	{
	  FD_ZERO (&writefds);
	  FD_SET (fd[1], &writefds);
	  if (n <= fd[1])
	    n = fd[1] + 1;
	}

      n = select (n, &readfds, (str != NULL ? &writefds : NULL), NULL, NULL);
      if (n < 0)
	error (EXIT_FAILURE, errno,
	       _("communication with sed subprocess failed"));
      if (str != NULL && FD_ISSET (fd[1], &writefds))
	goto try_write;
      if (FD_ISSET (fd[0], &readfds))
	goto try_read;
      /* How could select() return if none of the two descriptors is ready?  */
      abort ();
#endif

      /* Attempt to write.  */
#if HAVE_SELECT
    try_write:
#endif
      if (str != NULL)
	{
	  if (len > 0)
	    {
	      ssize_t nwritten = write (fd[1], str, len);
	      if (nwritten < 0 && !IS_EAGAIN (errno))
		error (EXIT_FAILURE, errno,
		       _("write to sed subprocess failed"));
	      if (nwritten > 0)
		{
		  str += nwritten;
		  len -= nwritten;
		}
	    }
	  else
	    {
	      /* Tell the child there is nothing more the parent will send.  */
	      close (fd[1]);
	      str = NULL;
	    }
	}
#if HAVE_SELECT
      continue;
#endif

      /* Attempt to read.  */
#if HAVE_SELECT
    try_read:
#endif
      if (length == allocated)
	{
	  allocated = allocated + (allocated >> 1);
	  result = xrealloc (result, allocated);
	}
      {
	ssize_t nread = read (fd[0], result + length, allocated - length);
	if (nread < 0 && !IS_EAGAIN (errno))
	  error (EXIT_FAILURE, errno, _("read from sed subprocess failed"));
	if (nread > 0)
	  length += nread;
	if (nread == 0 && str == NULL)
	  break;
      }
#if HAVE_SELECT
      continue;
#endif
    }

  close (fd[0]);

  /* Remove zombie process from process list.  */
  exitstatus = wait_subprocess (child, "sed");
  if (exitstatus != 0)
    error (EXIT_FAILURE, 0, _("sed subprocess terminated with exit code %d"),
	   exitstatus);

  /* Remove NUL bytes from result.  */
  {
    char *p = result;
    char *pend = result + length;

    for (; p < pend; p++)
      if (*p == '\0')
	{
	  char *q;

	  q = p;
	  for (; p < pend; p++)
	    if (*p != '\0')
	      *q++ = *p;
	  length = q - result;
	  break;
	}
  }

  *resultp = result;
  *lengthp = length;
}


static void
process_message (mp)
     message_ty *mp;
{
  const char *msgstr = mp->msgstr;
  size_t msgstr_len = mp->msgstr_len;
  size_t nsubstrings;
  char **substrings;
  size_t total_len;
  char *total_str;
  const char *p;
  char *q;
  size_t k;

  /* Count NUL delimited substrings.  */
  for (p = msgstr, nsubstrings = 0;
       p < msgstr + msgstr_len;
       p += strlen (p) + 1, nsubstrings++);

  /* Process each NUL delimited substring separately.  */
  substrings = (char **) xmalloc (nsubstrings * sizeof (char *));
  for (p = msgstr, k = 0, total_len = 0; k < nsubstrings; k++)
    {
      char *result;
      size_t length;

      process_string (p, strlen (p), &result, &length);
      result = xrealloc (result, length + 1);
      result[length] = '\0';
      substrings[k] = result;
      total_len += length + 1;
    }

  /* Concatenate the results, including the NUL after each.  */
  total_str = (char *) xmalloc (total_len);
  for (k = 0, q = total_str; k < nsubstrings; k++)
    {
      size_t length = strlen (substrings[k]);

      memcpy (q, substrings[k], length + 1);
      free (substrings[k]);
      q += length + 1;
    }
  free (substrings);

  mp->msgstr = total_str;
  mp->msgstr_len = total_len;
}


static void
process_message_list (mlp)
     message_list_ty *mlp;
{
  size_t j;

  for (j = 0; j < mlp->nitems; j++)
    process_message (mlp->item[j]);
}


static msgdomain_list_ty *
process_msgdomain_list (mdlp)
     msgdomain_list_ty *mdlp;
{
  size_t k;

  for (k = 0; k < mdlp->nitems; k++)
    process_message_list (mdlp->item[k]->messages);

  return mdlp;
}
